File: app/src/main/java/vandyke/siamobile/backend/GlobalPrefsListener.java
/*
 * Copyright (c) 2017 Nicholas van Dyke
 *
 * This file is subject to the terms and conditions defined in Licensing section of the file 'README.md'
 * included in this source code package. All rights are reserved, with the exception of what is specified there.
 */

package vandyke.siamobile.backend;

import android.content.*;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.BatteryManager;
import android.os.IBinder;
import vandyke.siamobile.backend.coldstorage.ColdStorageService;
import vandyke.siamobile.backend.siad.Siad;
import vandyke.siamobile.backend.siad.SiadMonitorService;
import vandyke.siamobile.misc.SiaMobileApplication;

public class GlobalPrefsListener implements SharedPreferences.OnSharedPreferenceChangeListener {

    private Context context;
    private ServiceConnection connection;
    private WalletMonitorService walletMonitorService;
    private boolean bound;

    public GlobalPrefsListener(Context context) {
        this.context = context;
        connection = new ServiceConnection() {
            public void onServiceConnected(ComponentName name, IBinder service) {
                walletMonitorService = ((WalletMonitorService.LocalBinder) service).getService();
                bound = true;
                walletMonitorService.postRefreshRunnable();
            }

            public void onServiceDisconnected(ComponentName name) {
                bound = false;
            }
        };
    }
    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        switch (key) {
            case "operationMode":
                if (sharedPreferences.getString("operationMode", "cold_storage").equals("cold_storage")) {
                    if (editor.putString("address", "localhost:9990").commit()) {
                        context.stopService(new Intent(context, SiadMonitorService.class));
                        context.startService(new Intent(context, ColdStorageService.class));
                    }
                } else if (sharedPreferences.getString("operationMode", "cold_storage").equals("remote_full_node")) {
                    if (editor.putString("address", sharedPreferences.getString("remoteAddress", "192.168.1.11:9980")).commit()) {
                        context.stopService(new Intent(context, ColdStorageService.class));
                        context.stopService(new Intent(context, SiadMonitorService.class));
                    }
                } else if (sharedPreferences.getString("operationMode", "cold_storage").equals("local_full_node")) {
                    if (editor.putString("address", "localhost:9980").commit()) {
                        context.stopService(new Intent(context, ColdStorageService.class));
                        context.startService(new Intent(context, SiadMonitorService.class));
                    }
                }
                new Thread(new Runnable() {
                    public void run() {
                        try {
                            Thread.sleep(1000); // sleep for 1 second to give whatever service/server was started time to start before querying it
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                        refreshWalletMonitorService();
                    }
                }).start();
                break;
            case "monitorRefreshInterval":
                if (!bound)
                    context.bindService(new Intent(context, WalletMonitorService.class), connection, Context.BIND_AUTO_CREATE);
                else
                    walletMonitorService.postRefreshRunnable();
                break;
            case "runLocalNodeOffWifi":
                ConnectivityManager connectivityManager =
                        (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
                NetworkInfo activeNetInfo = connectivityManager.getActiveNetworkInfo();

                if (activeNetInfo != null && activeNetInfo.getType() == ConnectivityManager.TYPE_WIFI
                        || SiaMobileApplication.prefs.getBoolean("runLocalNodeOffWifi", false)) {
                    context.startService(new Intent(context, Siad.class));
                } else {
                    context.stopService(new Intent(context, Siad.class));
                }
                break;
            case "localNodeMinBattery":
                Intent batteryStatus = context.registerReceiver(null, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
                int level = batteryStatus.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
                if (level >= Integer.parseInt(SiaMobileApplication.prefs.getString("localNodeMinBattery", "20")))
                    context.startService(new Intent(context, Siad.class));
                else
                    context.stopService(new Intent(context, Siad.class));
                break;
            case "remoteAddress":
                if (sharedPreferences.getString("operationMode", "cold_storage").equals("remote_full_node")) {
                    editor.putString("address", sharedPreferences.getString("remoteAddress", "192.168.1.11:9980"));
                    if (editor.commit())
                        refreshWalletMonitorService();
                }
                break;
        }
    }

    private void refreshWalletMonitorService() {
        if (!bound)
            context.bindService(new Intent(context, WalletMonitorService.class), connection, Context.BIND_AUTO_CREATE);
        else
            walletMonitorService.refreshAll();
    }
}

File: app/src/main/java/vandyke/siamobile/ui/settings/GlobalPrefsListener.kt
/*
 * Copyright (c) 2017 Nicholas van Dyke
 *
 * This file is subject to the terms and conditions defined in 'LICENSE.md'
 */

package vandyke.siamobile.ui.settings

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.SharedPreferences
import vandyke.siamobile.data.local.Prefs
import vandyke.siamobile.data.remote.SiaApi
import vandyke.siamobile.data.siad.SiadService

class GlobalPrefsListener(private val context: Context) : SharedPreferences.OnSharedPreferenceChangeListener {

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        when (key) {
            "operationMode" -> {
                when (Prefs.operationMode) {
                    "cold_storage" -> {
                        Prefs.address = "localhost:9990"
                        context.stopService(Intent(context, SiadService::class.java))
                    }
                    "remote_full_node" -> {
                        Prefs.address = Prefs.remoteAddress
                        context.stopService(Intent(context, SiadService::class.java))
                    }
                    "local_full_node" -> {
                        Prefs.address = "localhost:9980"
                        context.startService(Intent(context, SiadService::class.java))
                    }
                }
            }
            "runLocalNodeOffWifi" -> {
                if (Prefs.operationMode == "local_full_node") {
                    if (SiadService.isConnectionGood(context))
                        SiadService.singleAction(context, {
                            it.startSiad()
                        })
                    else
                        SiadService.singleAction(context, {
                            it.siadNotification("Stopped - not connected to WiFi")
                            it.stopSiad()
                        })
                }
            }
        // TODO: this is the cause of siad starting when wifi + battery requirements aren't met I think
            "localNodeMinBattery" -> {
                if (Prefs.operationMode == "local_full_node") {
                    val batteryStatus = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
                    if (SiadService.isBatteryGood(batteryStatus))
                        SiadService.singleAction(context, {
                            it.startSiad()
                        })
                    else
                        SiadService.singleAction(context, {
                            it.siadNotification("Stopped - battery is below ${Prefs.localNodeMinBattery}%")
                            it.stopSiad()
                        })
                }
            }
            "address" -> SiaApi.rebuildApi()
            "remoteAddress" -> if (Prefs.operationMode == "remote_full_node") {
                Prefs.address = Prefs.remoteAddress
            }
            "SiaNodeWakeLock" -> SiadService.getService(context).subscribe { service ->
                /* If Siad is already running then we must tell the service to acquire/release its wake lock
                   because normally it does so in start/stopSiad() */
                if (service.isSiadRunning) {
                    if (Prefs.SiaNodeWakeLock)
                        service.wakeLock.acquire()
                    else if (service.wakeLock.isHeld)
                        service.wakeLock.release()
                }
            }
        }
    }
}
